syntax = "proto3";

service ContinousLearning {
    rpc SetLabelName(LabelName) returns (Empty) {}

    rpc PutTrainImage(Image) returns (Empty) {}
    rpc GetTrainImage(ID) returns (Image) {}
    rpc DeleteTrainImage(Image) returns (Empty) {}
    rpc GetUnknownTrainImage(Filter) returns (stream Image) {}
    rpc SetTrainLabel(Label) returns (Empty) {}
    rpc GetTrainLabel(ID) returns (Label) {}

    rpc PutEvaluateImage(Image) returns (Empty) {}
    rpc GetEvaluateImage(ID) returns (Image) {}
    rpc DeleteEvaluateImage(Image) returns (Empty) {}
    rpc GetUnknownEvaluateImage(Filter) returns (stream Image) {}
    rpc SetEvaluateLabel(Label) returns (Empty) {}
    rpc GetEvaluateLabel(ID) returns (Label) {}

    rpc Evaluate(Empty) returns (Evaluation) {}
    rpc Predict(Image) returns (Prediction) {}
}

message Empty {}

message ID {
    bytes bytes = 1;
}

message LabelName {
    int32 label = 1;
    string name = 2;
}

message Image {
    ID id = 1;
    oneof image {
        string path = 2;
    }
}

message Filter {}

message Label {
    ID id = 1;
    oneof image {
        int32 label = 2;
        string name = 3;
    }
}

message Evaluation {}

message Prediction {}
